prediction_market_agent/agents/microchain_agent/blockchain/code_interpreter.py

```diff
@@ -39,8 +41,12 @@ def __init__(self, source_code: str, summarization_model: str = "gpt-4") -> None
         self.build_rag_chain()
 
     def build_retriever(self) -> None:
+        # 1 call per function
+        # no RAG - entire source code
+        # ChatOpenAI -> first call, here is the source code, 2nd call (with memory) give me a summary
+        # / of function a, 3rd call summary of function b, ...
         sol_splitter = RecursiveCharacterTextSplitter.from_language(
-            language=Language.SOL, chunk_size=128, chunk_overlap=0
+            language=Language.SOL, chunk_size=500, chunk_overlap=0
```

Comment:
Is this really solving the problem we discussed yesterday? Why not just give the LLM exact function definition instead of approximating it with RAG?

Reply:
Seems like it works nicely for wxDai, but it can still happen that by chunking, LLM will receive something irrelevant or only a part of the relevant function.

Anyway, it's good for me to leave it like this and get back to it  if it becomes a problem.

![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/1b52545d-984f-4275-85a7-fe805d0719bd)

Reply:
I would say this is a 80% solution. Agree to revisit this when problems arise.

Reply:
After discussion with @evangriffiths , I pivoted to the solution we also briefly discussed @kongzii - pass `source_code` to the LLM and ask for all summaries at once. It works.

```
pprint(cls)
defaultdict(<class 'list'>,
            {<AbiItemStateMutabilityEnum.NON_PAYABLE: 'nonpayable'>: [<class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Approve'>,
                                                                      <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Transfer'>,
                                                                      <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Transferfrom'>,
                                                                      <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Withdraw'>],
             <AbiItemStateMutabilityEnum.PAYABLE: 'payable'>: [<class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Deposit'>],
             <AbiItemStateMutabilityEnum.VIEW: 'view'>: [<class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Allowance'>,
                                                         <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Balanceof'>,
                                                         <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Decimals'>,
                                                         <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Name'>,
                                                         <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Symbol'>,
                                                         <class 'prediction_market_agent.agents.microchain_agent.blockchain.contract_class_converter.Wrappedxdaicontract_Totalsupply'>]})

```